subsys/bluetooth/controller/ll_sw/ull_llcp_common.c

```diff
@@ -0,0 +1,412 @@
+/*
+ * Copyright (c) 2020 Demant
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+
+#include <zephyr/types.h>
+
+#include <bluetooth/hci.h>
+#include <sys/byteorder.h>
+#include <sys/slist.h>
+#include <sys/util.h>
+
+#ifndef ULL_LLCP_UNITTEST
+#include "hal/ccm.h"
+
+#include "util/mem.h"
+#include "util/memq.h"
+
+#include "pdu.h"
+#include "ll.h"
+#include "ll_settings.h"
+
+#include "lll.h"
+#include "lll_conn.h"
+
+#include "ull_conn_types.h"
+#include "ull_tx_queue.h"
+#include "ull_llcp.h"
+#include "ull_llcp_internal.h"
+#endif
+
+#define BT_DBG_ENABLED IS_ENABLED(CONFIG_BT_DEBUG_HCI_DRIVER)
+#define LOG_MODULE_NAME bt_ctlr_ull_llcp
+#include "common/log.h"
+#include <soc.h>
+#include "hal/debug.h"
+
+/* LLCP Local Procedure Encryption FSM states */
+enum {
+	LP_COMMON_STATE_IDLE,
+	LP_COMMON_STATE_WAIT_TX,
+	LP_COMMON_STATE_WAIT_RX,
+	LP_COMMON_STATE_WAIT_NTF,
+};
+
+/* LLCP Local Procedure Common FSM events */
+enum {
+	/* Procedure run */
+	LP_COMMON_EVT_RUN,
+
+	/* Response recieved */
+	LP_COMMON_EVT_RESPONSE,
+
+	/* Reject response recieved */
+	LP_COMMON_EVT_REJECT,
+
+	/* Unknown response recieved */
+	LP_COMMON_EVT_UNKNOWN,
+
+	/* Instant collision detected */
+	LP_COMMON_EVT_COLLISION,
+};
+
+/* LLCP Remote Procedure Common FSM states */
+enum {
+	RP_COMMON_STATE_IDLE,
+	RP_COMMON_STATE_WAIT_RX,
+	RP_COMMON_STATE_WAIT_TX,
+	RP_COMMON_STATE_WAIT_NTF,
+};
+/* LLCP Remote Procedure Common FSM events */
+enum {
+	/* Procedure run */
+	RP_COMMON_EVT_RUN,
+
+	/* Request recieved */
+	RP_COMMON_EVT_REQUEST,
+};
+
+/*
+ * LLCP Local Procedure Common FSM
+ */
+
+static void lp_comm_tx(struct ull_cp_conn *conn, struct proc_ctx *ctx)
+{
+	struct node_tx *tx;
+	struct pdu_data *pdu;
+
+	/* Allocate tx node */
+	tx = tx_alloc();
+	LL_ASSERT(tx);
+
+	pdu = (struct pdu_data *)tx->pdu;
+
+	/* Encode LL Control PDU */
+	switch (ctx->proc) {
+	case PROC_VERSION_EXCHANGE:
+		ull_cp_priv_pdu_encode_version_ind(pdu);
+		ctx->rx_opcode = PDU_DATA_LLCTRL_TYPE_VERSION_IND;
+		break;
+	default:
+		/* Unknown procedure */
+		LL_ASSERT(0);
+	}
+
+	ctx->tx_opcode = pdu->llctrl.opcode;
+
+	/* Enqueue LL Control PDU towards LLL */
+	ull_tx_enqueue(conn, tx);
+}
+
+static void lp_comm_ntf(struct ull_cp_conn *conn, struct proc_ctx *ctx)
+{
+	struct node_rx_pdu *ntf;
+	struct pdu_data *pdu;
+
+	/* Allocate ntf node */
+	ntf = ntf_alloc();
+	LL_ASSERT(ntf);
+
+	ntf->hdr.type = NODE_RX_TYPE_DC_PDU;
+	pdu = (struct pdu_data *) ntf->pdu;
+
+	switch (ctx->proc) {
+	case PROC_VERSION_EXCHANGE:
+		ull_cp_priv_ntf_encode_version_ind(conn, pdu);
+		break;
+	default:
+		/* Unknown procedure */
+		LL_ASSERT(0);
+	}
+
+	/* Enqueue notification towards LL */
+	ll_rx_enqueue(ntf);
```

Comment:
The fuction declaration for ll_rx_enqueue is missing, this goes for the ull_llcp_enc.c and the ull_llcp_phy.c as well